virttest/qemu_devices/qbuses.py

```diff
@@ -650,6 +650,118 @@ def _increment_addr(self, addr, last_addr=None):
         return super(QPCIBus, self)._increment_addr(addr, last_addr=last_addr)
 
 
+class QPCIEBus(QPCIBus):
+    """
+    PCIE root bus representation (creates pcie-root-port for virtio pci
+    device by default.)
+    """
+
+    def __init__(self, busid, bus_type, aobject=None,
+                 root_type="pcie-root-port"):
+        """
+        :param busid: id of the bus (pcie.0)
+        :param bus_type: type of the bus (PCIE)
+        :param aobject: Related autotest object (pci.0)
+        :param root_type: root port type
+        """
+        super(QPCIEBus, self).__init__(busid, bus_type, aobject)
+        self.__root_ports = {}
+        self.__root_type = root_type
+
+    def is_direct_plug(self, device):
+        """
+        Justify if the device should be plug to pcie bus directly,
+        the device is not directly plugged only if it is a virtio
+        pci device and the device param direct_plug is "no".
+
+        :param device: the QBaseDevice object
+        :return: bool value for directly plug or not.
+        """
+        pcie_devices = ["virtio-net-pci", "virtio-blk-pci",
+                        "virtio-scsi-pci", "virtio-balloon-pci",
+                        "virtio-serial-pci", "virtio-rng-pci",
+                        "e1000e", "virtio-gpu-device"]
+        if device.get_param("driver") in pcie_devices and (
+                    device.get_param("direct_plug_pci", "no") == "no"):
+            return False
+
+        return True
+
+    def add_root_port(self, addr, root_port=None, multifunction="off"):
+        """
+        Add pcie root port to the bus and __root_ports list, assign free slot,
+        chassis, and addr for it according to current list.
+
+        :param addr: addr to add the root port
+        :param root_port: pcie-root-port QDevice object
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if not root_port:
+            root_port = self._add_root_port(addr, multifunction)
+        if _addr not in self.__root_ports:
+            bus = root_port.child_bus[0]
+            params = {'slot': _addr, 'chassis': _addr, 'addr': "0x%x" % _addr}
+            for key, value in params.iteritems():
+                root_port.set_param(key, value)
+            self.__root_ports["0x%x" % _addr] = bus
+        return root_port
+
+    def _add_root_port(self, addr, multifunction="off"):
+        """
+        Generate pcie-root-port QDevice object with certain addr and id
+
+        :param addr: addr to add the root port
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if _addr not in self.__root_ports:
```

Comment:
the same here, either you shoud store `self.__root_ports` as `_addr`s, or you need to convert it to `0x%x` everywhere (I'd go with the first, but not a strong opinion, though)

Reply:
again, not necessary as `insert` won't allow inserting 2 devices into the same address.